grakn-graql/src/main/java/ai/grakn/graql/internal/analytics/KCoreVertexProgram.java

```diff
@@ -0,0 +1,260 @@
+    public KCoreVertexProgram() {
+    }
+
+    public KCoreVertexProgram(int kValue) {
+        this.k = kValue;
+        this.persistentProperties.put(K, kValue);
+    }
+
+    @Override
+    public void storeState(final Configuration configuration) {
+        super.storeState(configuration);
+        persistentProperties.put(K, k);
+    }
+
+    @Override
+    public void loadState(final Graph graph, final Configuration configuration) {
+        super.loadState(graph, configuration);
+        k = (int) persistentProperties.get(K);
+    }
+
+    @Override
+    public Set<VertexComputeKey> getVertexComputeKeys() {
+        return VERTEX_COMPUTE_KEYS;
+    }
+
+    @Override
+    public Set<MemoryComputeKey> getMemoryComputeKeys() {
+        return MEMORY_COMPUTE_KEYS;
+    }
+
+    @Override
+    public void setup(final Memory memory) {
+        LOGGER.debug("KCoreVertexProgram Started !!!!!!!!");
+
+        // K_CORE_STABLE is true by fault, and we reset it after each odd iteration.
+        memory.set(K_CORE_STABLE, false);
+        memory.set(K_CORE_EXIST, false);
+        memory.set(VOTE_TO_HALT, true);
+        memory.set(CONNECTED_COMPONENT_STARTED, false);
+    }
+
+    @Override
+    public void safeExecute(final Vertex vertex, Messenger<String> messenger, final Memory memory) {
+        String id;
+        switch (memory.getIteration()) {
+            case 0:
+                sendMessage(messenger, EMPTY_MESSAGE);
+                break;
+
+            case 1: // get degree first, as degree must >= k
+                if ((vertex.label().equals(Schema.BaseType.ENTITY.name()) ||
+                        vertex.label().equals(Schema.BaseType.ATTRIBUTE.name())) &&
+                        Iterators.size(messenger.receiveMessages()) >= k) {
+                    id = vertex.value(Schema.VertexProperty.ID.name());
+                    vertex.property(CLUSTER_LABEL, id);
+                    memory.add(K_CORE_EXIST, true);
+
+                    // send ids from now on, as we want to count connected entities, not relationships
+                    sendMessage(messenger, id);
+                }
+                break;
+
+            default:
+                if (memory.<Boolean>get(CONNECTED_COMPONENT_STARTED)) {
+                    if (messenger.receiveMessages().hasNext()) {
+                        if (vertex.property(CLUSTER_LABEL).isPresent()) {
+                            updateClusterLabel(vertex, messenger, memory);
+                        } else if (vertex.label().equals(Schema.BaseType.RELATIONSHIP.name())) {
+                            relayClusterLabel(messenger, memory);
+                        }
+                    }
+                } else {
+                    if (memory.getIteration() % 2 == 0) {
+                        relayOrSaveMessages(vertex, messenger);
+                    } else {
+                        updateEntityAndAttribute(vertex, messenger, memory);
+                    }
```

Comment:
I think I get this: we do this odd-even thing so we relay messages "through" relationships? I'd add a comment explaining it.